mlc_llm/relax_model/llama.py

```diff
@@ -211,30 +234,52 @@ def forward(
         past_key_value: Tuple[relax.Expr],
         attention_mask: Optional[relax.Expr] = None,
     ) -> Tuple[relax.Expr, Optional[relax.Expr], Optional[Tuple[relax.Expr]]]:
-        from tvm.relax.op import astype, matmul, maximum, permute_dims, reshape, squeeze
+        from tvm.relax.op import (
+            astype,
+            matmul,
+            maximum,
+            permute_dims,
+            reshape,
+            split,
+            squeeze,
+        )
         from tvm.relax.op.nn import softmax
 
         bsz, q_len, _ = hidden_states.struct_info.shape
         assert bsz == 1, "Only support batch size 1 at this moment."
 
-        query_states = nn.emit(
-            reshape(
-                self.q_proj(hidden_states),
-                (bsz, q_len, self.num_heads, self.head_dim),
-            ),
-        )
-        key_states = nn.emit(
-            reshape(
-                self.k_proj(hidden_states),
-                (bsz, q_len, self.num_heads, self.head_dim),
-            ),
-        )
-        value_states = nn.emit(
-            reshape(
-                self.v_proj(hidden_states),
-                (bsz, q_len, self.num_heads, self.head_dim),
-            ),
-        )
+        if self.combine_qkv:
+            # (bsz * q_len, 3 * self.num_heads, self.head_dim)
+            qkv_states = nn.emit(
+                reshape(
+                    self.query_key_value_proj(hidden_states),
+                    (bsz, q_len, 3 * self.num_heads, self.head_dim),
+                ),
+            )
+            qkv_states = nn.emit(split(qkv_states, indices_or_sections=3, axis=-2))
```

Comment:
For compatibility purpose, can you make the result of this transformation identical to what https://github.com/apache/tvm/blob/unity/src/relax/transform/combine_parallel_matmul.cc does? 

In particular I don't understand why we need reshape here.

Reply:
Do you mean “splitting right after matmul”? Yes sure will check out the pass later. The reshape here is in accommodation of KV cache and it can definitely go after the split.

Reply:
yes exactly. Yeah I have reshapes after split in my mod using the combining pass.

Reply:
Okay updated now.